### DISCUSSION TITLE
ESPNow.init() documentation

### DISCUSSION THREAD

**Discussion by mmliam** [micropython/micropython#11509] [General]
I've searched the various Micropython ESPNow references and have not seen a specific descriptive reference for ESPNow.init(). From some experimentation it appears to include ESPNow.active(true); what other functions does it perform? Can someone provide a documentation link.

Thanks, MM

**Comment by jimmo** [comment]
@mmliam There is no `ESPNow.init()`. Not quite sure what you're asking sorry?

**Reply by mmliam** [reply]
There is, Glenn referenced it on the old micropython.org site and I've use it in code, without ESPNow.active(true) and it works. It certainly doesn't throw any errors.
Of course the "ESPNow" is a pseudo object.

Found this reference by Glenn toward the bottom:
https://forum.micropython.org/viewtopic.php?t=9177&start=20

SAMPLE:

> Code: [Select all](https://forum.micropython.org/viewtopic.php?t=9177&start=20#)

    e = espnow.ESPNow()
    e.init(100)

    seemed to work better with a value in init() Also, need to work out how to get the msgs 

It *should* not be necessary to provide an argument to init(), but on the ESP8266 you can provide the read buffer size as an argument - as there is NO config() call on the 8266 (because the code size requirements are too tight) (see my docs). 100 bytes seems small for the buffer size.

**Reply by jimmo** [reply]
There were some minor API changes between the earlier versions and what was ultimately merged into upstream. The docs and examples reflect the final API: https://docs.micropython.org/en/latest/library/espnow.html

> Of course the "ESPNow" is a pseudo object.

Not sure what you mean... It's a real class (albeit with a singleton base class, which I guess is a bit odd). https://github.com/micropython/micropython/blob/master/ports/esp32/modules/espnow.py

**Reply by mmliam** [reply]
All the text below SAMPLE are from a discussion between dave & Glenn.

The .init() method does seem to be an outlier that's why I've been trying to get more information. Apparently it still functional as I've been using it. It must call the .active() method because I haven't used that but my ESPnow code works. I don't like using functions I don't understand. As you say it may be functional, but obsolete. Wish I knew what it does, or was supposed to do - lol.

**Reply by jimmo** [reply]
> Apparently it still functional as I've been using it.

If you're using official firmware from micropython.org (i.e. any nightly build since early May) then I don't know how this is possible. There is no `init` method on the `ESPNow` class, or the `ESPNowBase` it inherits from.

**Reply by glenn20** [reply]
`ESPNow.init()/ESPNow.deinit()` was deprecated some time ago in favour of `ESPNow.active(True)/ESPNow.active(False)`. I believe I made a post about that some time ago in the forum (but I undertstand that could have been easily missed). This change makes it behave more like other drivers in micropython (eg. WLAN()) and has the advantage of smaller code size (One C method instead of two).

There have been some other changes to the API as part of the merge process into micropython. The history of this can be seen in the (long) discussion on PR #6515 (eg. changes to `ESPNow.on_recv() -> ESPNow.irq()`  and some structural changes to simplify and reduce the code size of the C module. ie. the C module has been pared back and renamed `_espnow`. The ESPNow user API is now presented through the ESPNow class in `ports/esp32/modules/espnow.py` which inherits from the `ESPNowBaseClass` in the `_espnow` C module. I hope that those changes have happened mostly transparently to the users of the API.

I regret any disruption from API changes, but some changes were inevitable as part of the merge process (and have improved the code, and consistency with other micropython APIs).

**Reply by glenn20** [reply]
Oh - and the upshot is - if you are using an older image, the current docs for `ESPNow.active(True)` will apply to `ESPNow.init()`.

**Reply by mmliam** [reply]
Thanks Glenn & jimmo, the code I was using ESPNow.init() is several years old; I was just making some changes. After reviewing the software I ran across the instruction but could find any documentation. I suspected it may have been replaced by ESPNow.active(True), and was equivalent. I'll update the software to the current format.
